### PR TITLE
Add Synchronous Metric Instruments SDK

### PR DESCRIPTION
**Pull request by ankit-bhargava**
This PR implements classes described in the Metric Instruments API ([PR #161](https://github.com/open-telemetry/opentelemetry-cpp/pull/161)).  These classes, unlike the no-op implementations included with the API, are capable of actually collecting telemetry data and sit at the beginning of the Metrics data pipeline as they capture data from instrumented applications.  Instruments employ Aggregators (pending approval in [PR #178](https://github.com/open-telemetry/opentelemetry-cpp/pull/178) ) to combine the updates they receive into meaningful values.  **The PR will not compile without these files.**

A test suite which validates each instrument’s ability to record values in single-threaded and concurrent scenarios is included.  Functions necessary for the metrics pipeline as a whole such as reference counting are tested as well.

This PR is part of a series of PRs to add a full implementation of metrics API and SDK support. 

Note: We templated instruments to support a wider array of scalar types.  Templating prevents us from using out of line declarations as we normally would.  As a result, we left all implementation code in the header file.  If there are any suggestions to remedy this we will gladly make the change.

### CODE REVIEW

File: sdk/include/opentelemetry/sdk/metrics/instrument.h [hunk 1]
```diff
@@ -0,0 +1,230 @@
...
+    
+    SynchronousInstrument(nostd::string_view name,
+                          nostd::string_view description,
+                          nostd::string_view unit,
+                          bool enabled,
+                          metrics_api::InstrumentKind kind):Instrument(name,description,unit,enabled,kind)
+    {}
+    
+    /**
+     * Returns a Bound Instrument associated with the specified labels.         * Multiples requests
```
Comment (reyang) [inline sdk/include/opentelemetry/sdk/metrics/instrument.h:146]: Is this a format error?

File: sdk/include/opentelemetry/sdk/metrics/instrument.h [hunk 2]
```diff
@@ -0,0 +1,230 @@
...
+        default:
+            throw std::invalid_argument("Labels must be strings");
+            break;
+    }
+};
+
+// Utility function which converts maps to strings for better performance
+inline std::string mapToString(const std::map<std::string,std::string> & conv){
+    std::stringstream ss;
+    ss <<"{ ";
```
Comment (reyang) [inline sdk/include/opentelemetry/sdk/metrics/instrument.h:195]: There are several places in this PR where the code is not well-formatted (e.g. spacing, indentation, etc.)

Comment (ankit-bhargava) [reply sdk/include/opentelemetry/sdk/metrics/instrument.h:195]: Sorry about that, for some reason these weren't an issue in my IDE.  I'll be sure to go over them carefully and take care of all the issues.

Comment (ankit-bhargava) [reply sdk/include/opentelemetry/sdk/metrics/instrument.h]: went back over and tried to clean up most of these

**Comment by reyang** [comment]
@ankit-bhargava would you rebase and resolve the conflicts?

**Comment by ankit-bhargava** [comment]
@reyang I had some issues rebasing so please let me know if anything looks off.  Otherwise, this should be ready to go!

**Comment by ankit-bhargava** [comment]
@reyang Recently rebased, think this can be merged.

**Pull request merged**